kotlin/show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgisruntime/sample/showdevicelocationusingindoorpositioning/MainActivity.kt

```diff
@@ -0,0 +1,404 @@
+        val field = fields.find { it.name.equals("DateCreated", ignoreCase = true) || it.name.equals("Date_Created", ignoreCase = true) }
+        return field?.name
+    }
+
+    /**
+     * Retrieves the PathwaysTable
+     */
+    private fun getPathwaysTable(): ArcGISFeatureTable? {
+        try {
+            val pathwaysFeatureLayer = mapView.map.operationalLayers.firstOrNull { it.name.equals("Pathways") } as? FeatureLayer
+            return pathwaysFeatureLayer?.featureTable as? ArcGISFeatureTable
+        } catch (e: Exception) {
+            // if pathways table not found in map's operationalLayers
+            val message = "PathwaysTable not found"
+            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+            Log.e(TAG, message)
+            return null
+        }
+    }
+
+    /**
+     * Invokes when the app is closed or LocationDataSource is stopped.
+     */
+    private fun stopLocationDisplay() {
+        mapView.locationDisplay.stop()
+        mapView.locationDisplay.locationDataSource.apply {
+            removeStatusChangedListener(this@MainActivity)
+            removeLocationChangedListener(this@MainActivity)
+        }
+    }
+
+    /**
+     * Sets up the location listeners, the navigation mode, and display's the devices location as a blue dot
+     */
+    private fun startLocationDisplay() {
+        val locationDisplay: LocationDisplay = mapView.locationDisplay.apply {
+            autoPanMode = LocationDisplay.AutoPanMode.NAVIGATION
+            locationDataSource = mIndoorsLocationDataSource
+        }
+        // these listeners will receive location, heading and status updates from the location data source.
+        locationDisplay.locationDataSource.apply {
+            addStatusChangedListener(this@MainActivity)
+            addLocationChangedListener(this@MainActivity)
+        }
+        // asynchronously start of the location display,
+        // which will in-turn start IndoorsLocationDataSource to start receiving IPS updates.
+        locationDisplay.startAsync()
+    }
+
+    /**
+     * Handles the data received on a [locationChangedEvent]
+     * like currentFloor, positionSource, transmitterCount, networkCount and horizontalAccuracy
+     */
+    override fun locationChanged(locationChangedEvent: LocationDataSource.LocationChangedEvent?) {
+        // get the location properties of the LocationDataSource
+        val locationProperties = locationChangedEvent?.location?.additionalSourceProperties
+        if(locationProperties == null){
+            Toast.makeText(this, "LocationDataSource does not have any property-fields", Toast.LENGTH_LONG).show()
+            Log.e(TAG, "LocationDataSource does not have any property-fields")
+            return
+        }
+        // retrieve information about the location of the device
+        val floor =
+            if (locationProperties["floor"] != null)
+                locationProperties["floor"].toString()
+            else ""
```

Comment:
can use ?: elvis operators here and below

for example `val floor = (locationProperties["floor"] ?: "").toString()`